base/compiler/typeinfer.jl

```diff
@@ -105,44 +133,31 @@ reset_timings()
     parent_timer = _timings[end]
     accum_time = stop_time - parent_timer.cur_start_time
 
-    # Add in accum_time ("modify" the immutable struct)
+    # Add in accum_time
     @inbounds begin
-        _timings[end] = Timing(
-            parent_timer.mi_info,
-            parent_timer.start_time,
-            parent_timer.cur_start_time,
-            parent_timer.time + accum_time,
-            parent_timer.children,
-            parent_timer.bt,
-        )
+        _timings[end].time += accum_time
```

Comment:
Should this be an atomic increment?

Reply:
it doesn't need to be, since it's essentially a thread-local object since (as of julia 1.8) this is guarded by the codgen lock.

Indeed it's not thread safe starting in 1.9, that's what https://github.com/JuliaLang/julia/pull/48051 is for.
We were going to merge these separately so that we can backport this to 1.8 in our own build.

Reply:
The right strategy then is to have them be separate commits in the same PR.

Reply:
Hm, so you think we should move that PR's changes into this one before merging this one? 🤔 

I just squashed this one down to one commit to make that easier; can also make this into an atomic

Reply:
Hm, I tried this just now but I'm not allowed to `using Base.Threads: Atomic, atomic_add!` in the compiler, since it's not defined yet 🤔

Reply:
i'm fine with doing it in one PR. But note that this is already broken without this PR, and so doing it in two incremental steps didn't seem worse than doing it in one step, and that way we could merge this one which is already pretty much done...

But we can merge them both into here if that's easier, yeah.

Reply:
Okay, just reviewed the commit in the thread safety PR again (https://github.com/JuliaLang/julia/pull/48051/commits/18f30c8b6da7aff311e5a5dfc17799cc6c67bde8), and I don't think that it's finished. I don't think we tested it at all, and we certainly haven't tested it in our codebase yet.

I would not feel comfortable ramming that in yet, BUT i would really like us to merge this PR to close this loop and prevent it from bitrotting.

Since this doesn't make the thread safety any _worse_, let's merge this PR now, and then we can come back to the thread safety discussion in https://github.com/JuliaLang/julia/pull/48051.

Reply:
> Maybe we can't use atomics and need to rely on a lock?

`@atomic` not `Base.Threads.Atomic` to mark a field as atomic.

Reply:
Okay, i merged in https://github.com/JuliaLang/julia/pull/48051. We're testing this now in our 1.9 branch at RAI, and assuming it looks okay then i'll feel good about moving forward with this.

I guess it's also got its own tests in this PR, so it's probably fine too.

Reply:
Thanks 🙏